### PR TITLE
feature: Unified API

### PR DESCRIPTION
**Pull request by marandaneto**
## :loudspeaker: Type of change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
feature: Unified API

## :bulb: Motivation and Context
SDK follows the Sentry convention for SDKs
https://develop.sentry.dev/sdk/unified-api/

## :green_heart: How did you test it?
running samples and unit/integration tests

## :pencil: Checklist

- [X] I reviewed submitted code
- [X] I added tests to verify changes
- [X] All tests passing
- [ ] No breaking changes

### CODE REVIEW

**Comment by bruno-garcia** [comment]
Make sure not to squash this PR 👍

File: dart/lib/src/sentry_options.dart [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  /// sentry.{language}.{platform}/{version} eg sentry.java.android/2.0.0 would be a valid case
+  String sentryClientName;
+
+  /// This function is called with an SDK specific event object and can return a modified event
+  /// object or nothing to skip reporting the event
+  BeforeSendCallback beforeSendCallback;
+
+  /// This function is called with an SDK specific breadcrumb object before the breadcrumb is added
+  /// to the scope. When nothing is returned from the function, the breadcrumb is dropped
+  BeforeBreadcrumbCallback beforeBreadcrumbCallback;
```
Comment (bruno-garcia) [inline dart/lib/src/sentry_options.dart]:
```suggestion
  BeforeBreadcrumbCallback beforeBreadcrumb;
```

File: dart/lib/src/sentry_options.dart [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  List<String> get inAppIncludes => List.unmodifiable(_inAppIncludes);
+
+  Transport _transport = NoOpTransport();
+
+  Transport get transport => _transport;
+
+  set transport(Transport transport) =>
+      _transport = transport ?? NoOpTransport();
+
+  // TODO: transportGate, connectionTimeoutMillis, readTimeoutMillis, hostnameVerifier, sslSocketFactory, proxy
```
Comment (bruno-garcia) [inline dart/lib/src/sentry_options.dart]: We add on demand? 
```suggestion
```

File: dart/lib/src/sentry_options.dart [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  /// The server name used in the Sentry messages.
+  String serverName;
+
+  Sdk _sdk = Sdk(name: sdkName, version: sdkVersion);
+
+  /// Sdk object that contains the Sentry Client Name and its version
+  Sdk get sdk => _sdk;
+
+  set sdk(Sdk sdk) {
+    _sdk = sdk ?? _sdk;
```
Comment (bruno-garcia) [inline dart/lib/src/sentry_options.dart:146]:
```suggestion
    _sdk =?? sdk;
```

File: dart/lib/src/transport/transport.dart [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      compressPayload: _options.compressPayload,
+    );
+
+    final response = await _options.httpClient.post(
+      dsn.postUri,
+      headers: _credentialBuilder.configure(_headers),
+      body: body,
+    );
+
+    if (response.statusCode != 200) {
```
Comment (bruno-garcia) [inline dart/lib/src/transport/transport.dart]: Highly recommend a `log` `warning` here.

even a `if (options.level == debug) { log whole server response }`

File: dart/lib/src/transport/transport.dart [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      return SentryId.empty();
+    }
+
+    final eventId = json.decode(response.body)['id'];
```
Comment (bruno-garcia) [inline dart/lib/src/transport/transport.dart]: Does this work on the envelope endpoint?

File: dart/test/stack_trace_test.dart [hunk 6]
```diff
@@ -79,23 +79,25 @@ void main() {
...
-          'colno': 3,
-          'in_app': true,
-          'filename': 'test.dart'
-        },
-      ]);
-    });
+//       expect(encodeStackTrace('''
+// #0      baz (file:///pathto/test.dart:50:3)
+// #1      bar (file:///pathto/secret.dart:46:9)
+//       ''', stackFrameFilter: filter), [
```
Comment (bruno-garcia) [inline dart/test/stack_trace_test.dart]: these will be deleted before we merge? Are they no longer relevant?

File: dart/lib/src/sentry_options.dart (hunk 2 above)
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart]: I'll create a card for it

File: dart/lib/src/transport/transport.dart (hunk 5 above)
Comment (marandaneto) [reply dart/lib/src/transport/transport.dart]: we don't use the envelope endpoint, but /store, the envelope is part of the next version

File: dart/test/stack_trace_test.dart (hunk 6 above)
Comment (marandaneto) [reply dart/test/stack_trace_test.dart]: should be removed, this has been replaced by beforeSend

File: dart/lib/src/sentry_client.dart [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+  Random _random;
+
+  /// Reports an [event] to Sentry.io.
+  Future<SentryId> captureEvent(
+    SentryEvent event, {
+    Scope scope,
+    dynamic hint,
+  }) async {
+    final emptyFuture = Future.value(SentryId.empty());
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart]: the Future do you mean? cus `SentryId.empty()` already returns a static field

File: dart/lib/src/sentry_options.dart [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  set compressPayload(bool compressPayload) =>
+      _compressPayload = compressPayload ?? _compressPayload;
+
+  Client _httpClient = Client();
+
+  /// If [httpClient] is provided, it is used instead of the default client to
+  /// make HTTP calls to Sentry.io. This is useful in tests.
+  Client get httpClient => _httpClient;
+
+  set httpClient(Client httpClient) => _httpClient = httpClient ?? _httpClient;
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart:38]: you probably mean: `??=` which is:

> Assign the value only if the variable is null

if `_httpClient` is non-null, then it'll never be overwritten

File: dart/lib/src/sentry_options.dart [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+  /// If [clock] is provided, it is used to get time instead of the system
+  /// clock. This is useful in tests. Should be an implementation of [ClockProvider].
+  ClockProvider _clock = getUtcDateTime;
+
+  ClockProvider get clock => _clock;
+
+  set clock(ClockProvider clock) => _clock = clock ?? _clock;
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart:46]: https://github.com/getsentry/sentry-dart/pull/117#discussion_r513041363

File: dart/lib/src/sentry_options.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart:146]: https://github.com/getsentry/sentry-dart/pull/117#discussion_r513041363

**Comment by marandaneto** [comment]
a few TODOs:

https://github.com/getsentry/sentry-dart/projects/1#card-48213380
https://github.com/getsentry/sentry-dart/projects/1#card-47082476
https://github.com/getsentry/sentry-dart/projects/1#card-47082234

File: dart/lib/src/sentry_options.dart (hunk 2 above)
Comment (marandaneto) [reply dart/lib/src/sentry_options.dart]: https://github.com/getsentry/sentry-dart/projects/1#card-48213567

File: dart/lib/src/transport/transport.dart (hunk 5 above)
Comment (marandaneto) [reply dart/lib/src/transport/transport.dart]: https://github.com/getsentry/sentry-dart/projects/1#card-48213613

**Comment by marandaneto** [comment]
Draft of the docs https://github.com/getsentry/sentry-docs/pull/2599

File: dart/lib/src/sentry_options.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply dart/lib/src/sentry_options.dart:38]: yeah `??=`


Right, I ignore me

File: dart/lib/src/sentry_client.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply dart/lib/src/sentry_client.dart]: yeah just the future

File: dart/lib/src/transport/transport.dart (hunk 4 above)
Comment (bruno-garcia) [reply dart/lib/src/transport/transport.dart]: Maybe log error btw

**Review by rxlabz — APPROVED** [review]
LGTM

**Pull request merged**